### PR TITLE
Design proposal: kpt live -- Inventory to ResourceGroup

### PR DESCRIPTION
**Pull request by yuwenma**
This PR is a design proposal on enabling kpt live apply to deploy from STDIN and migrate inventory from Kptfile to a standalone ResourceGroup object/file.

### CODE REVIEW

File: docs/design-docs/01-kpt-live-stdin.md
Comment (mikebz) [inline docs/design-docs/01-kpt-live-stdin.md:31]: Does that not already exist?

File: docs/design-docs/01-kpt-live-stdin.md
Comment (mikebz) [inline docs/design-docs/01-kpt-live-stdin.md:54]: we might want to say something about the alpha (v.39) of kpt.  We used to have a separate inventory there.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (mikebz) [inline docs/design-docs/01-kpt-live-stdin.md:92]: minor nit - can we use the MD syntax here?

File: docs/design-docs/01-kpt-live-stdin.md
Comment (mikebz) [inline docs/design-docs/01-kpt-live-stdin.md:78]: should these be two separate flags --resourcegroup-path or --resourcegroup-file?

File: docs/design-docs/01-kpt-live-stdin.md
Comment (mikebz) [inline docs/design-docs/01-kpt-live-stdin.md:129]: it's unlikely that Helm will spit out a ResourceGroup, right?

File: docs/design-docs/01-kpt-live-stdin.md
Comment (mikebz) [inline docs/design-docs/01-kpt-live-stdin.md:138]: I would consider deprecating it in the future and talking about the timeline

**Review by mikebz — COMMENTED** [review]
Looks very good.  Just a couple of minor questions.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (rquitales) [reply docs/design-docs/01-kpt-live-stdin.md:31]: Yes, `kpt live apply -` [already exists](https://github.com/GoogleContainerTools/kpt/blob/3b42daa41102fa83bce07bd82a72edcd691868d6/internal/cmdapply/cmdapply.go#L52) and [works with STDIN](https://github.com/GoogleContainerTools/kpt/blob/3b42daa41102fa83bce07bd82a72edcd691868d6/pkg/live/load.go#L72). It does require a Kptfile with Inventory information in the stream.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (phanimarupaka) [reply docs/design-docs/01-kpt-live-stdin.md:78]: I would use rg short-form instead.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (phanimarupaka) [inline docs/design-docs/01-kpt-live-stdin.md:84]: What if the package has inventory-template.yaml?

File: docs/design-docs/01-kpt-live-stdin.md
Comment (phanimarupaka) [inline docs/design-docs/01-kpt-live-stdin.md:110]: How does this work with `kpt fn source <DIR> | kpt live apply -`. Is `resourcegroup.yaml` treated as meta-resource or normal resource ?

File: docs/design-docs/01-kpt-live-stdin.md
Comment (phanimarupaka) [inline docs/design-docs/01-kpt-live-stdin.md:80]: So the resourcegroup.yaml file can be anywhere in the directory tree of the package? Assume a package hierarchy of mysql in wordpress package. What does it mean to do `kpt live init wordpress/mysql` and `kpt fn source wordpress | kpt live apply -` ? Should we have a restriction to have the file in the top level directory of the package ?

File: docs/design-docs/01-kpt-live-stdin.md
Comment (marshall007) [inline docs/design-docs/01-kpt-live-stdin.md:43]: Will `kpt live init` still accept arguments for `--inventory-id`, `--name`, and `--namespace`?

> Default value has prefix “inventory-” with 8 digit numbers. E.g. “inventory-02014045”

I believe `--name` currently defaults to the package name. Is the implication here that a random suffix will now be used? I think it would be desirable for `kpt live init . --inventory-id="..."` to continue producing deterministic output.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (droot) [inline docs/design-docs/01-kpt-live-stdin.md:39]: What is rationale for appending this 8 digit number ? (when one specifiy it from the commandline then the number will still be appended or not ?) Also, is this related to the `inventory-id` ?

File: docs/design-docs/01-kpt-live-stdin.md
Comment (droot) [reply docs/design-docs/01-kpt-live-stdin.md:43]: We should continue to accept those arguments. I also have the same question about the name and inventory-id.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (droot) [inline docs/design-docs/01-kpt-live-stdin.md:58]: s/v1betaX/v1

Note the CLI is in beta but the schema is v1.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (droot) [reply docs/design-docs/01-kpt-live-stdin.md:80]: `kpt fn source wordpress | kpt live apply -` would result in failure because the `fn source wordpress` output will contain two resourcegroup CRs, one from `mysql` and another from `wordress` and as the 1) says, it will report the failure.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (droot) [inline docs/design-docs/01-kpt-live-stdin.md:88]: We avoid the term `hydrate` in the user/design docs, instead pl. use `render` everywhere.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (droot) [reply docs/design-docs/01-kpt-live-stdin.md:110]: Thats a good point. Quick thought suggest - since info was part of Kptfile earlier that was meta-resource, should probably be considered as meta-resource, need to think more.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (droot) [reply docs/design-docs/01-kpt-live-stdin.md:78]: I see `resourcegroup-file` as the flag name in other parts of the design.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (droot) [inline docs/design-docs/01-kpt-live-stdin.md:73]: We should may be include a paragraph explaining how `kpt live apply ` populates this resource with the inventory list etc. and mention never to run `kubectl apply rg.yaml` to the cluster because it will remove the info populated by the `live apply`. And we should also document, how someone can recover if they accidently run `kubectl apply` with rg.yaml ?

**Review by droot — COMMENTED** [review]
Thanks for submitting the design here in the repo. I have some minor comments.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:31]: updated.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:39]: this is just random number, I guess it's to avoid name duplication. If user provides the "name", it will use what user specified. 
Not related to inventory-id.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:43]: It will, it will generate a standalone ResourceGroup with the provided flags. See `Simplify the Inventory`, inventory-id can be skipped.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:54]: That has been well handled by kpt (kpt live migrate) and seems irrelevant to this doc?

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:58]: Good catch and Thank you!

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:78]: Good catch and thank you all! changed to `--resourcegroup-file` and +1 to --rg as short-form

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:84]: The functionality of existing `kpt live migrate` is still valid, it is just extended to move the inventory to a standalone file. You can understand it as providing two migration paths
1.  inventory-template.yaml --> resourcegroup.yaml (used to be Kptfile)
2. kptfile --> resourcegroup.yaml

File: docs/design-docs/01-kpt-live-stdin.md
Comment (mikebz) [reply docs/design-docs/01-kpt-live-stdin.md:54]: sure that's fine.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:84]: oh, I see what you refer to here. Do you think we can ignore the inventory-template.yaml use case? I feel like it is more rare than Kptfile and potentially could cause more overhead/latency since it will analyze the configMaps.  
Or, we can add that condition check if users file issues. WDYT? 

@rquitales FYI, inventory-template.yaml is an earlier version and is what the current `kpt live migrate` solves. If we can have this corner case well handled, that would be super cool.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:88]: Oh,  I cannot agree more on avoiding the term "hydrate" in user/design docs. It will need another 1000+ words to explain if people ask for clarification.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:92]: Sure. done

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:138]: yeah, Sunil Mortan and I have discussed about it. And we want to have more signals before moving forward. Deprecating the inventory might be aggressive without user data support.

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:79]: Defer to @rquitales. We can have the user guide once the feature is complete. You might find more cases want to talk about at that time :)

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:157]: qq: `kpt fn source` outputs a ResourceList CR, thus `kpt live apply -` won't deploy the k8s resources in ResourceList.items  but the actual ResourceList (and require the CRD to be preinstalled), right? Why do we suggest `kpt fn source <DIR> | kpt live apply -` here? @phanimarupaka is there any issues/discussions about it? 

We expect to have `kpt live apply` from STDIN be neutral to the hydration tool, so it does not assume the inputs are of any special kpt kind

File: docs/design-docs/01-kpt-live-stdin.md
Comment (yuwenma) [reply docs/design-docs/01-kpt-live-stdin.md:176]: Why not using `kpt live init` to generate a resourceGroup? and then they can use helm + kpt deploy, is that not wonderful?  😄

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (mikebz) [inline docs/design-docs/01-live_Inventory-to-ResourceGroup.md:60]: curious what is the advantage of using a label as opposed to a field?

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (mikebz) [inline docs/design-docs/01-live_Inventory-to-ResourceGroup.md:76]: very nice that name and namespace are now in the usual and customary places.

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (mikebz) [inline docs/design-docs/01-live_Inventory-to-ResourceGroup.md:105]: does this have to be this long?  Is there another place for --resourcegroup where it wouldn't be a file?

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (mikebz) [inline docs/design-docs/01-live_Inventory-to-ResourceGroup.md:140]: suggestion is use the example with a conventional file name.  Most of the users love to copy and paste.  Don't be surprised if we will start getting a bunch of CUSTOM_RG.yaml files out there.  I'd prefer not to have that.

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (mikebz) [inline docs/design-docs/01-live_Inventory-to-ResourceGroup.md:228]: I don't follow this.  This is almost exactly the same data.  What's different?

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (yuwenma) [reply docs/design-docs/01-live_Inventory-to-ResourceGroup.md:60]: In general, labels are preferred in two aspects (especially compared to annotations): one is the query efficiency and the other is for its identification characteristic. Thus it sounds reasonable to use label here, but I don't know the exact reason for cli-utils.

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (mikebz) [reply docs/design-docs/01-live_Inventory-to-ResourceGroup.md:105]: Did I miss an answer here? 😄

**Comment by mikebz** [comment]
@yuwenma seemed like a couple of comments were resolved without answers.  If there is a blanket strategy to resolve it - let me know.  I think it might be good for people to know why those alternatives were provided or not considered.

**Comment by mikebz** [comment]
🚥 ping....

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (droot) [reply docs/design-docs/01-live_Inventory-to-ResourceGroup.md:105]: we can use `--rg-file` ? `--resourcegroup` gives the impression that it is expecting name of the resource group.

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (droot) [inline docs/design-docs/01-live_Inventory-to-ResourceGroup.md:108]: I don't think CLI library `cobra` (that is used by kpt) supports two letter short form. I have suggested to use `--rg-file` above.

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (droot) [inline docs/design-docs/01-live_Inventory-to-ResourceGroup.md:111]: s/invectory/inventory

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (droot) [inline docs/design-docs/01-live_Inventory-to-ResourceGroup.md:117]: s/invectory/inventory

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (droot) [inline docs/design-docs/01-live_Inventory-to-ResourceGroup.md:123]: s/invectory/inventory

**Review by droot — COMMENTED** [review]
@yuwenma Proposal is looking in good shape. Two open questions:

- suggestion on using short name for the commandline flag. `--rg-file` is what I have suggested.
- I think something we overlooked about the interplay of inventory information with functions such as `namespace` and `ensure-name-substring`. High level question to answer is: should `kpt fn render or kpt fn eval kpt fn source` should ignore the inventory resource or leave it to the functions to ignore it. If we want to maintain the current behavior, then `kpt fn` should ignore the inventory resource when invoking functions.

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (yuwenma) [reply docs/design-docs/01-live_Inventory-to-ResourceGroup.md:111]: Good catch, thank you.

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (yuwenma) [reply docs/design-docs/01-live_Inventory-to-ResourceGroup.md:117]: done

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (yuwenma) [reply docs/design-docs/01-live_Inventory-to-ResourceGroup.md:123]: done

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (yuwenma) [reply docs/design-docs/01-live_Inventory-to-ResourceGroup.md:108]: Took a quick look at cobra, it uses the pflag which accepts one-letter shorthand for flags (Maybe I misread the code). https://github.com/spf13/pflag/blob/master/flag.go#L173

**Comment by yuwenma** [comment]
@droot  Updated to '--rg-file`

File: docs/design-docs/01-kpt-live-stdin.md
Comment (phanimarupaka) [reply docs/design-docs/01-kpt-live-stdin.md]: @yuwenma We need a way to pass inventory/RG information to `kpt live apply` through STDIN. But treating ResourceGroup as a meta resource will make functions modify it which is catastrophic. So we should treat ResourceGroup resource as a special resource and not modify it as part of functions and not include it in `kpt fn source`. In future, if there are use-cases which users come up with, we can include a new flag `--include-rg` for `kpt fn` commands. So for now, I am good with not supporting `kpt fn source <DIR> | kpt live apply -`.

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (phanimarupaka) [reply docs/design-docs/01-live_Inventory-to-ResourceGroup.md:108]: Yes. Shorthands are one letter only.

**Review by phanimarupaka — APPROVED** [review]

**Comment by phanimarupaka** [comment]
@yuwenma Thanks for the doc. Approved it.

@droot We can merge it once you approve.

File: docs/design-docs/01-live_Inventory-to-ResourceGroup.md
Comment (droot) [reply docs/design-docs/01-live_Inventory-to-ResourceGroup.md:108]: Yeah, that confirms that short form `--rg` wont work.

**Comment by droot** [comment]
@yuwenma I had a discussion with @phanimarupaka  today about how to treat inventory resource wrt to `kpt fn` and `kpt pkg` commands.  Pl. take a look at the summary of the discussion in the last comment. Can you pl. ack the proposal in the comment (below) and would be even better if we can capture this summary in the doc itself somewhere because it impacts the implementation.

https://github.com/GoogleContainerTools/kpt/pull/2576#discussion_r760548405

**Comment by phanimarupaka** [comment]
@droot @yuwenma I added the commit about the rg file interaction with fn commands.

**Review by droot — APPROVED** [review]

**Comment by phanimarupaka** [comment]
Merging this PR. Thanks for your contribution @yuwenma. @rquitales Please feel free to start implementing this and let us know if you have any questions.

cc @mikebz

**Pull request merged**
